### PR TITLE
[nrf noup] Add tfm read service to nRF SoC Flash driver

### PR DESCRIPTION
**Pull request by sigvartmh**
This adds the functionality to read memory which are residing in the
secure part of flash from a non-secure application when building with
TF-M.

Currently we check if the address space is not inside the non-secure
applapplication address space. If the address space is outside we do a
request to TF-M so that the secure application can either return success
or fail if the address space is not allowed to be read.

Signed-off-by: Sigvart Hovland <sigvart.hovland@nordicsemi.no>

### CODE REVIEW

**Review by nvlsianpu — COMMENTED** [review]
We don't have any Kconfig propert which told us that application is build for running along with TFM. Is that right?

File: drivers/flash/soc_flash_nrf.c [hunk 1]
```diff
@@ -153,10 +158,20 @@ static int flash_nrf_read(const struct device *dev, off_t addr,
...
-	if (addr < PM_APP_ADDRESS) {
+#if CONFIG_ARM_NONSECURE_FIRMWARE && USE_PARTITION_MANAGER
+	if ((addr < PM_APP_ADDRESS) || (addr > (PM_APP_ADDRESS + PM_APP_SIZE))) {
+#if CONFIG_SPM && CONFIG_SPM_SECURE_SERVICES
 		return spm_request_read(data, addr, len);
+#else
+		uint32_t err = 0;
+		enum tfm_platform_err_t plt_err;
+		plt_err = tfm_platform_mem_read(&data, addr, len, &err);
+		if (plt_err != TFM_PLATFORM_ERR_SUCCESS || err != 0) {
```
[the remark this answers is not in the document]
Comment (sigvartmh) [reply drivers/flash/soc_flash_nrf.c:169]: Do you want to remove dTFM_PLATFORM_ERR_SUCCESS or?

So this checks if `err != 0` which is the return value of the `mem_read` function while the function returns the `plt_err` which is the error code reported if there is some violation in the TF-M platform/world.

**Comment by sigvartmh** [comment]
I could probably hide this behind `CONFIG_BUILD_WITH_TFM`

File: drivers/flash/soc_flash_nrf.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (nvlsianpu) [reply drivers/flash/soc_flash_nrf.c:169]: `if (plt_err != TFM_PLATFORM_ERR_SUCCESS)` would be enough assuming that TFM_PLATFORM_ERR_SUCCESS is equal 0.

[the remark this answers is not in the document]
Comment (sigvartmh) [reply drivers/flash/soc_flash_nrf.c:169]: You can get `err` even if ´TFM_PLATFORM_ERR_SUCCESS`.

**Review by hakonfam — APPROVED** [review]
Please add a jira ticket to the commit message

File: drivers/flash/soc_flash_nrf.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (nvlsianpu) [reply drivers/flash/soc_flash_nrf.c:170]: what an oversight of mine

**Review by nvlsianpu — APPROVED** [review]

**Comment by mglettig** [comment]
Could you update NRF-SDKs west.yml manifest file to point to the merge commit that integrates this PR into master of sdk-zephyr? I'm interesting in testing those changes.
Kind regards,
Michael

**Comment by tejlmand** [comment]
> Zephyr is frozen unfortunately.

Zephyr is frozen, but sdk-nrf is not, so the manifest can be updated by sending a PR with new sha here: https://github.com/nrfconnect/sdk-nrf

**Comment by carlescufi** [comment]
Why is this a `[noup]`? This doesn't feel like it should be one, @SebastianBoe and @joerchan.

**Comment by SebastianBoe** [comment]
I see there are references to SPM, which is NCS-only.

Perhaps this commit could be split in two.

One commit relevant for SPM, which will be deprecated and then we can drop the commit. So still technically a noup I believe.

And one commit relevant for TF-M.

@joerchan , @sigvartmh ?

**Comment by joerchan** [comment]
@SebastianBoe @carlescufi See https://github.com/nrfconnect/sdk-nrf/pull/6461#issuecomment-1010966141
Also tracked in NCSDK-13242

**Comment by SebastianBoe** [comment]
How can this be toup, isn't SPM NCS-only ?

**Comment by SebastianBoe** [comment]
toup is banned BTW, we would need to create a PR and change to fromlist.

**Comment by carlescufi** [comment]
> toup is banned BTW, we would need to create a PR and change to fromlist.

Exactly, toup hasn't been allowed for a while now, so @sigvartmh and @joerchan can we please find a better solution? This needs to be reverted in the meantime.

**Comment by joerchan** [comment]
That seems drastic, toup should be avoided I agree, but for the situation where the commit needs to be reworked before it can be submitted upstream it should be the right one to use.

**Comment by SebastianBoe** [comment]
Discussed offline with Joakim.

Working on a better solution. Should be trivial so I would prefer not to revert unless I have to task switch.

**Comment by carlescufi** [comment]
> That seems drastic, toup should be avoided I agree, but for the situation where the commit needs to be reworked before it can be submitted upstream it should be the right one to use.

Yes, but people were abusing that sauce tag, as it often happens with these sorts of "catch-all" waivers, so we decided to disallow it. `wip` is allowed but only for temporary work before it is either sent upstream or converted into a proper `noup`

**Comment by carlescufi** [comment]
> Discussed offline with Joakim.
> 
> Working on a better solution. Should be trivial so I would prefer not to revert unless I have to task switch.

OK, as long as this is resolved before you force-push to the main branch in this repo once the upmerge PR is ready.

**Comment by SebastianBoe** [comment]
Reverted as you suggested @carlescufi as I had to task-switch.

Will re-apply if we find that it is needed for the upmerge but I have no reason to believe it is needed for the upmerge.

@sigvartmh : Action on you to re-apply this after resolving the issues mentioned in: https://github.com/nrfconnect/sdk-zephyr/pull/689

**Comment by SebastianBoe** [comment]
@carlescufi : Please give a +1 in #689

**Pull request merged**